File: solutions/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = 0

        ch_set = set()

        l = 0
        for r, ch in enumerate(s):
            if ch in ch_set:
                while (s[l] != ch):
                    ch_set.remove(s[l])
                    l+=1
                l+=1
            else:
                ch_set.add(ch)
                res = max(res, r-l+1)

        return res

File: solutions/Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.ts
/**
STRATEGY:
    sliding window
    iterate with 2 pointers,
        one for beginning 
        one for end
    use a set to track which chars are in the window
    when a duplicate is found, 
        iterate left pointer until duplicate is removed
 */

function lengthOfLongestSubstring(s: string): number {
    let longest: number = 1;
    const set: Set<string> = new Set<string>();

    let [l,r] = [0,1];
    set.add(s[0]);
    while (r < s.length) {
        // first verify window and adjust if needed
        if (set.has(s[r])) {
            // move left pointer to duplicate, and remove from set as we go
            while (s[l] !== s[r]) {
                set.delete(s[l]);
                l++;
            }
            // l now at duplicate
            l++;
        }
        else set.add(s[r]);

        // window is now valid
        longest = Math.max(longest, r-l+1);

        // iterate r
        r++;
    }

    return longest;
};